File: yumi/edge/template/yumi_tools/typescript/yumiSetup.ts
/**
 * Yumi Edge — TypeScript tool registration
 *
 * Register your functions with `agent.register()` and call `initYumi()`
 * from your app entry point.
 *
 *     import { initYumi } from "./yumi_tools/typescript/yumiSetup";
 *     initYumi();
 *
 * Quick test: ``npx tsx yumiSetup.ts`` (from yumi_tools/typescript/)
 *
 * Requires: npm install (from the yumi_tools/typescript/yumi_sdk directory)
 */

import path from "node:path";
import { pathToFileURL } from "node:url";

import { YumiAgent } from "./yumi_sdk/src";

// ── Connection (edit here, or set in .env) ──

const YUMI_CONNECTION_CODE = "yumi-lan_..."; // paste from `yumi --server`
const YUMI_EDGE_NAME = "My Node App";

export function initYumi(): YumiAgent {
  const agent = new YumiAgent({
    connectionCode: YUMI_CONNECTION_CODE,
    edgeName: YUMI_EDGE_NAME,
  });

  // ── Register tools: name + description + parameters + handler ──

  // agent.register({
  //   name: "jump",
  //   description: "Make the character jump",
  //   parameters: [
  //     { name: "height", type: "number", description: "Jump height in meters" },
  //   ],
  //   handler: async (args) => {
  //     const height = args.number("height") ?? 1.0;
  //     return `Jumped ${height} meters`;
  //   },
  // });

  // Dangerous tools: user confirms in the Yumi web UI or `yumi --chat` (not on device):
  // agent.register({
  //   name: "delete_all",
  //   description: "Delete all data",
  //   requireConfirmation: true,
  //   handler: async () => "Deleted everything",
  // });

  agent.runInBackground();
  return agent;
}

/** True when this file is executed directly (e.g. `npx tsx yumiSetup.ts`). */
function isDirectRun(): boolean {
  if (typeof process === "undefined" || !process.argv[1]) {
    return false;
  }
  return import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
}

if (isDirectRun()) {
  initYumi();
  console.log("Yumi edge running (yumiSetup as entry). Press Ctrl+C to stop.");
}

File: yumi/edge/template/yumi_tools/cpp/yumi_setup.cpp
/**
 * Yumi Edge — C++ tool registration
 *
 * Register your functions with agent.registerTool() and call initYumi()
 * from your main program entry point.
 *
 * Build: add the YumiSDK directory via CMake add_subdirectory(), then
 * link with target_link_libraries(your_app PRIVATE yumi_sdk).
 *
 * See README.md for full setup instructions.
 */

#include <yumi/yumi_agent.hpp>
#include <string>
#include <utility>

// ── Connection (edit here, or set in .env) ──

static const char* YUMI_CONNECTION_CODE = "yumi-lan_...";  // from `yumi --server`
static const char* YUMI_EDGE_NAME = "My C++ App";

yumi::YumiAgent* initYumi() {
    auto* agent = new yumi::YumiAgent(YUMI_CONNECTION_CODE, YUMI_EDGE_NAME);

    // ── Register tools: name + description + parameters + handler ──

    // yumi::RegisterOptions jump;
    // jump.name = "jump";
    // jump.description = "Make the character jump";
    // jump.parameters = {
    //     {"height", "number", "Jump height in meters"},
    // };
    // jump.handler = [](const yumi::ToolArguments& args) -> std::string {
    //     double h = args.number("height").value_or(1.0);
    //     return "Jumped " + std::to_string(h) + " meters";
    // };
    // agent->registerTool(std::move(jump));

    // Dangerous tools: user confirms in the Yumi web UI or `yumi --chat` (not on device):
    // yumi::RegisterOptions deleteAll;
    // deleteAll.name = "delete_all";
    // deleteAll.description = "Delete all data";
    // deleteAll.requireConfirmation = true;
    // deleteAll.handler = [](const yumi::ToolArguments&) -> std::string {
    //     return "Deleted everything";
    // };
    // agent->registerTool(std::move(deleteAll));
    //
    // Read-only tools can opt in to proactive messaging context:
    // yumi::RegisterOptions status;
    // status.name = "get_status";
    // status.description = "Read current app status";
    // status.allowProactive = true;
    // status.proactiveContext = true;
    // status.handler = [](const yumi::ToolArguments&) -> std::string {
    //     return "ok";
    // };
    // agent->registerTool(std::move(status));

    agent->runInBackground();
    return agent;
}

// Example main (remove or replace with your own):
// int main() {
//     auto* agent = initYumi();
//     // ... your main loop ...
//     // agent->stop();
//     // delete agent;
//     return 0;
// }
